cpp/src/dictionary/dictionary_factories.cu

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/dictionary/dictionary_factories.hpp>
+
+
+namespace cudf
+{
+
+std::unique_ptr<column> make_dictionary_column( column_view const& keys_column,
+                                                column_view const& indices_column,
+                                                rmm::mr::device_memory_resource* mr,
+                                                cudaStream_t stream)
+{
+    CUDF_EXPECTS( !keys_column.has_nulls(), "keys column must not have nulls" );
+    if( keys_column.size()==0 )
+        return make_empty_column( data_type{DICTIONARY32} );
+    CUDF_EXPECTS( indices_column.type().id()==cudf::type_id::INT32, "indices column must be INT32" );
+
+    auto keys_copy = std::make_unique<column>( keys_column, stream, mr );
+    auto indices_copy = std::make_unique<column>( indices_column, stream, mr);
+    rmm::device_buffer null_mask;
+    auto null_count = indices_column.null_count();
+    if( null_count )
+    {
+        indices_copy->set_null_mask(null_mask,0);
+        null_mask = copy_bitmask(indices_column,stream,mr);
```

Comment:
Hm, there's an extra copy of the `indices_column` null mask happening here.

It gets copied above when doing:
```
    auto indices_copy = std::make_unique<column>(indices_column, stream, mr);
```
and then copied again here.

I see what you're trying to do here, and I'm trying to think of a clean way to do it. Basically, we want to copy the data and null mask from `indices_column` into two different columns.

One option is to do something like:
```c++
auto indices_copy = std::make_unique<column>(indices_column, stream, mr);
auto indices_contents = indices_copy.release();
auto indices = std::make_unique<column>(indices_column.type(), indices_column.size(), std::move(*indices_contents.data));

...
return std::make_unique<column>(data_type{DICTIONARY32}, indices_column.size(), rmm::device_buffer{}, std::move(*indices_contents.null_mask));
```